### PR TITLE
Get rid of pkg_resources

### PR DESCRIPTION
**Pull request by flying-sheep**
# Description

`pkg_resources` is a deprecated, extremely slow-to-import package that only exists when `setuptools` is installed.

Instead of adding a runtime dependency on `setuptools`, one should just use `importlib.metadata` and `importlib.resources` from the stdlib.

See here for performance benefits: https://github.com/mtkennerly/dunamai/issues/28

# TODO

- [x] [Changelog](CHANGELOG.md) entry

### CODE REVIEW

File: api/python/quilt3/session.py [hunk 1]
```diff
@@ -23,33 +26,33 @@
...
 AUTH_PATH = BASE_PATH / 'auth.json'
 CREDENTIALS_PATH = BASE_PATH / 'credentials.json'
-VERSION = pkg_resources.require('quilt3')[0].version
+VERSION = metadata.version('quilt3')
 
 
 def _load_auth():
     if AUTH_PATH.exists():
-        with open(AUTH_PATH, encoding='utf-8') as fd:
+        with AUTH_PATH.open(encoding='utf-8') as fd:
```
Comment (dimaryaz) [inline api/python/quilt3/session.py:34]: Is this the preferred way? I can't find anything about which `open` is better.

In any case, if this way is better, could you do it in a separate PR, just cause it's not related to `pkg_resources`?

File: gendocs/build.py [hunk 2]
```diff
@@ -34,10 +36,9 @@ def gen_walkthrough_doc():
 
 def install_pydocmd():
     try:
-        pydocmd_dist = pkg_resources.get_distribution('pydoc-markdown')  # install name, not module name
-        version = pydocmd_dist.version
-    except pkg_resources.DistributionNotFound:
-        version = ''
+        version = metadata.version('pydoc-markdown')  # install name, not module name
+    except metadata.PackageNotFoundError:
+        version = None
```
Comment (dimaryaz) [inline gendocs/build.py:41]: This will break the `version.endswith` below. Change it back to `None`, or change the check?

**Review by dimaryaz — CHANGES_REQUESTED** [review]
Thanks for the fix! A couple comments below...

File: gendocs/build.py (hunk 2 above)
Comment (flying-sheep) [reply gendocs/build.py:41]: done, check without whitespace changes: https://github.com/quiltdata/quilt/pull/2994/files?w=1

**Review by dimaryaz — APPROVED** [review]

File: docs/CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,7 @@
 ## Python API
 * [Fixed] Fix check to determine if a file is a tempfile in Windows with Python 3.8+ ([#2900](https://github.com/quiltdata/quilt/pull/2900))
 * [Changed] Disable upload optimization for objects with SSE-KMS ([#2790](https://github.com/quiltdata/quilt/pull/2790))
+* [Fixed] Sped up import and got rid of undeclared runtime dependency on `setuptools` ([#2994](https://github.com/quiltdata/quilt/pull/2994))
```
Comment (nl0) [inline docs/CHANGELOG.md:12]:
```suggestion
* [Fixed] Speed up import and get rid of undeclared runtime dependency on `setuptools` ([#2994](https://github.com/quiltdata/quilt/pull/2994))
```

**Review by nl0 — APPROVED** [review]

**Review by sir-sigurd — APPROVED** [review]

**Comment by sir-sigurd** [comment]
Thanks!

**Pull request merged**